sympy/solvers/ode.py

```diff
@@ -5930,7 +5898,24 @@ def _remove_coefficient(expr, x):
         # then they will need to be multiplied by sufficient x to make them so.
         # This function DOES NOT do that (it doesn't even look at the
         # homogeneous equation).
-        retdict['trialset'] = _get_trial_set(expr, x)
+        temp_set = set([])
+        cartesian_prod = set([])
+        if expr.is_Add:
+            for i in expr.args:
+                act = _get_trial_set(i,x)
+                if eq_homogeneous is not S.Zero:
+                    f = func.func
+                    while any([expand(sub_func_doit(eq_homogeneous, f(x), ts)).is_zero for ts in act]):
+                        act = {x*ts for ts in act}
+                temp_set = temp_set.union(act)
+        else:
+            cartesian_prod = _get_trial_set(expr, x)
+            if eq_homogeneous is not S.Zero:
+                f = func.func
+                while any([expand(sub_func_doit(eq_homogeneous, f(x), ts)).is_zero for ts in cartesian_prod]):
```

Comment:
What you have written is
```
any([f(x) for x in y])
```
This creates a list after completely iterating over `y` and calling `f` several times. Afterwards the list is passed to `any`. If you write this as
```
any(f(x) for x in y)
```
then a generator will be used. If any of the x is true the loop exits (short-circuits).

It is almost always preferable to use a generator rather than a list with something like `any`.

Reply:
Also you should make a short local function with a good name for `expand(sub_func_doit(eq_homogeneous, f(x), ts)).is_zero` perhaps `is_homogeneous_solution`.

Reply:
should make this inside `_undetermined_coefficients_match`?